Implement gmlm

IsabelCanoA opened the pull request:
Including the Generalized Multilinear Model

TODOS:

- [ ] We should implement "tempered" regularization a la Eric Lock's paper. When I spoke with him, he noted that it helped the algorithm find global minima more often.

IsabelCanoA committed:
Adding GMLM

IsabelCanoA committed:
exporting gmlm libraries

IsabelCanoA committed:
Faster Array

dahong67 commented on src/api-gmlm.jl:
```diff
@@ -0,0 +1,122 @@
+        end
+
+        # ---- update U-grad ----
+        wi = khatrirao(reverse(V)...)' * vec(X[i])
```
The `khatrirao` in this line only needs to be computed once outside the for loop and can be reused across the samples. Ditto for the `khatrirao` in `zi`.

Can you move it out and do some benchmarking to see if it makes any difference? I'm curious how much time is being spent here.

dahong67 reviewed (commented):
Just took a look through `gmlm_grad!` - great job on this!

I think we may be able to get some further speed-ups. Take a look at the comments and let me know how it goes!

dahong67 committed:
Add missing import

IsabelCanoA committed:
removing docstring in Array, it is covered by general Array docstring

IsabelCanoA committed:
Merge branch 'faster-cpd' of https://github.com/IsabelCanoA/GCPDecompositions.jl into faster-cpd

IsabelCanoA committed:
Creating a function copy! with buffers

IsabelCanoA committed:
Formating the code

dahong67 committed:
Drop MKL

IsabelCanoA committed:
Avoid coying in Array

IsabelCanoA committed:
Merge branch 'faster-cpd' into adding-gmlm

Adding faster-cpd implementation in this branch

IsabelCanoA committed:
moving khatrirao operation out of the for loop

IsabelCanoA committed:
formatting the code

dahong67 committed:
Merge branch 'faster-cpd' of github.com:IsabelCanoA/GCPDecompositions.jl into pr/IsabelCanoA/147

dahong67 committed:
Run formatter

dahong67 committed:
Reorder methods

dahong67 committed:
Simplify `Array` method

dahong67 committed:
Restrict `copy!` to `Array` types

Makes it a bit easier to reason about.
Can always relax later!

dahong67 committed:
Use `ndims` consistently in `copy!`

dahong67 committed:
Rename `copy!` args to match docstring

dahong67 committed:
Add specific tests for `AbstractArray` and `copy!`

dahong67 committed:
Check axes of inputs to `copy!`

dahong67 committed:
Move fast path to top

dahong67 committed:
Simplify code for collecting scaled matrices

dahong67 committed:
Reorder dims in tests so that the optimal split varies

Allows us to better exercise the code in `copy!`.
In particular, to make sure the buffers are right-sized.

dahong67 committed:
Simplify code for computing the products

dahong67 committed:
Experiment with skipping khatri-rao when not needed

dahong67 committed:
Revert "Experiment with skipping khatri-rao when not needed"

This reverts commit 2a03e05df234c16a0046adabb3bc892c8ec6e2ea.

Adds some complexity and may not have much gain in "typical" settings
since the larger products often dominate the cost.
Didn't see much difference in the following benchmark:
```julia
julia> M = CPD(ones(2), rand.((300,400,500),2));
julia> Y = Array{Float64}(undef, size(M));
julia> buffers = GCPDecompositions.create_copy_buffers(Array{Float64}(undef,size(M)), M);
julia> @benchmark copy!($Y, $M; buffers=$buffers)
```

dahong67 committed:
Use `transpose` since we shouldn't conjugate

`permutedims` may actually be most "correct"
since `transpose` is recursive, but I'm not sure
that hits the correct code-paths for BLAS.

dahong67 committed:
Split point only needs to go up to N-1

dahong67 committed:
No buffers needed for the N=1 fast path

dahong67 committed:
Drop `find_split_point`

IMHO, the one-liner `argmin` is much more readable
and probably not meaningfully slower (esp. given that
this step is unlikely to be a bottleneck).

dahong67 committed:
Tidy up and simplify `create_copy_buffers`

dahong67 committed:
Reorder methods

dahong67 committed:
Reorder imports

IsabelCanoA committed:
Merge branch 'faster-cpd' into adding-gmlm

IsabelCanoA committed:
replacing  contract with copy!

IsabelCanoA committed:
adding contratct with copy!

IsabelCanoA committed:
removing the prints of lbfgsb

IsabelCanoA committed:
using contract_cp_copy! instead of contract!

IsabelCanoA committed:
reoving the contract function from gmlm objective

IsabelCanoA committed:
Merge branch 'avoid-contract' into adding-gmlm

IsabelCanoA committed:
cleaning the functions removing the function contract!, using copy! instead
